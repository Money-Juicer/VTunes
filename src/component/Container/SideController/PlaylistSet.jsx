import React,{useEffect, useState} from "react";
import styles from "../../../styles/PlaylistSet.module.css";
import ScrollList from "../../common/ScrollList/ScrollList";
import PlaylistItem from "./PlaylistItem";

const PlaylistSet = ({
  isPlMenuClick, 
  onPlMenuClick, 
  listOfPlaylist, 
  onLoadAllPlaylists, 
  onSelectedPlaylist, 
  onAddPlaylist, 
  onDeletePlaylist,
  onIsCurrentPlaylistViewed,
  inputRef,
}) => {
  const [isDeleteClick, setIsDeleteClick] = useState(false);
  const [isAddClick, setIsAddClick] = useState(false);
  const [userInput, setUserInput] = useState("");
  const [isAlert, setIsAlert] = useState(false);
  
  useEffect(()=>{
    onLoadAllPlaylists();//listOfPlaylist를 로드한다
  // eslint-disable-next-line react-hooks/exhaustive-deps
  },[isPlMenuClick])
  
  const handleDeleteClick = ()=>{
    setIsDeleteClick(prev=>!prev);
  }
  const handleAddClick = () =>{
    setIsAddClick(prev=>!prev);
  }
  const handleAddPlaylist = () => {//playlist json 파일 추가
    setIsAddClick(false);
  
    const playlistExists = listOfPlaylist&&listOfPlaylist.some(playlist => playlist.name === userInput);
  
    if (playlistExists || userInput === "") {
      setUserInput("");
      setIsAlert(true);
      setIsAddClick(true);
    } else {
      onAddPlaylist({ name: userInput, list: [] }); //redux store에 추가
      setUserInput("");
    }
  };
  
  const handleDeletePlaylist = ()=>{
    setIsDeleteClick(false);
  }
  const handleInputChange = (event) => { //adder-modal에서 입력시
    setUserInput(event.target.value);
  };
  const handleKeyDown = (event) => {
    if (event.key === "Enter") {
      handleAddPlaylist();
    }
  };

  const buttonStyle = {
    fontSize: "20px",
    height: "70%",
    width: "35%",
  };

  return (

    <div className={styles["playlist-set-wrapper"]}>
      {isAddClick&&(//new playlist클릭시 뜨는 창 : adder-modal
        <div className={styles["adder-modal"]}>
          <span>{isAlert ? "Invalid Playlist Name" : "New Playlist"}</span>
          <div className={styles["search-engine"]} cypress-testid="playlist_name_input">
            <input ref = {inputRef}
              type="text"
              placeholder="       Enter the name for the New Playlist :)"
              value={userInput}
              onChange={handleInputChange}
              onKeyDown={handleKeyDown}
            />
          </div>
          <div className={styles["adder-modal-button-area"]}>
            <button className={styles["confirm-button"]} onClick={handleAddPlaylist} cypress-testid="confirm_button">Confirm</button>
            <button className={styles["cancel-button"]} onClick={() => { setIsAddClick(false); setIsAlert(false); }}>Cancel</button>
          </div>
        </div>
      )}
      <div className={styles["header"]} />{/*본문*/}
      <div className={styles["playlist-adder"]}>
        <button style={buttonStyle} onClick={handleAddClick} cypress-testid="new_playlist_button">New Playlist</button>
        <button style={buttonStyle} onClick={handleDeleteClick}>Delete Playlist</button>
      </div>
      <div className={styles["playlist-set"]}>
        <ScrollList>
        {listOfPlaylist &&
          Array.isArray(listOfPlaylist) &&
          listOfPlaylist
            .filter(playlist => playlist.name !== "현재재생목록")
            .map((playlist, index) => (
              <div className={styles["playlist-wrapper"]} key={index}>
                <PlaylistItem 
                  buttonFlag={isDeleteClick}
                  playlistData={playlist}
                  onPlMenuClick={onPlMenuClick}//플레이리스트 선택하면 sideContainerContents보여주는 용도
                  onSelectedPlaylist={onSelectedPlaylist}
                  onDeletePlaylist={onDeletePlaylist}
                  onIsCurrentPlaylistViewed={onIsCurrentPlaylistViewed}
                />
              </div>
            ))
        }
        </ScrollList>
      </div>
    </div>
  );
};

export default PlaylistSet;
