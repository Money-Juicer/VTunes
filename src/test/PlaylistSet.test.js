import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import ReduxProviderWrapper from './ReduxProviderWrapper';
import PlaylistSet from '../component/Container/SideController/PlaylistSet';
import { Provider } from 'react-redux';

// props Mock
const mockProps = {
    isPlMenuClick: false,
    onPlMenuClick: jest.fn(),
    listOfPlaylist: [
        { name: 'Playlist 1', list: [] },
        { name: 'Playlist 2', list: [] },
    ],
    onLoadAllPlaylists: jest.fn(),
    onSelectedPlaylist: jest.fn(),
    onAddPlaylist: jest.fn(),
    onDeletePlaylist: jest.fn(),
    onIsCurrentPlaylistViewed: jest.fn(),
};

describe('PlaylistSet', () => {
    it('renders PlaylistSet component', async () => {
        render(
        <ReduxProviderWrapper>
            <PlaylistSet {...mockProps} />
        </ReduxProviderWrapper>
        );
        expect(screen).toMatchSnapshot();

        // New Playlist
        const newPlaylistBtn = screen.getByText('New Playlist');
        expect(newPlaylistBtn).toBeInTheDocument();
        expect(newPlaylistBtn).toMatchSnapshot();
        fireEvent.click(newPlaylistBtn);

        const confirmBtn = screen.getByText('Confirm');
        expect(confirmBtn).toBeInTheDocument();
        expect(confirmBtn).toMatchSnapshot();
        
        const scrollList = screen.getByTestId('scrollList');
        expect(scrollList).toBeInTheDocument();
        expect(scrollList).toMatchSnapshot();
     
        // Delete Playlist
        const deletePlaylistBtn = screen.getByText('Delete Playlist');
        expect(deletePlaylistBtn).toBeInTheDocument();
        expect(deletePlaylistBtn).toMatchSnapshot();
    });

    
});

