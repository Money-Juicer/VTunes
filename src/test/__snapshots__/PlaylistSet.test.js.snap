// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlaylistSet renders PlaylistSet component 1`] = `
Object {
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "logTestingPlaygroundURL": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
}
`;

exports[`PlaylistSet renders PlaylistSet component 2`] = `
<button
  style="font-size: 20px; height: 70%; width: 35%;"
>
  New Playlist
</button>
`;

exports[`PlaylistSet renders PlaylistSet component 3`] = `
<button
  class="confirm-button"
>
  Confirm
</button>
`;

exports[`PlaylistSet renders PlaylistSet component 4`] = `
<div
  class="scroll-list-container"
  data-testid="scrollList"
>
  <div
    class="scroll-list"
  >
    <div
      class="playlist-wrapper"
    >
      <div
        class="playlist-wrapper"
      >
        <div
          class="delete-button-area"
        />
        <div
          class="playlist-info"
        >
          <div
            class="name"
          >
            Playlist 1
          </div>
          <div
            class="songs-count"
          >
            0
            ㆍSongs
          </div>
        </div>
        <div
          class="playlist-enter"
        >
          <img
            alt="playlistEnter"
            src="plEnter.png"
          />
        </div>
      </div>
    </div>
    <div
      class="playlist-wrapper"
    >
      <div
        class="playlist-wrapper"
      >
        <div
          class="delete-button-area"
        />
        <div
          class="playlist-info"
        >
          <div
            class="name"
          >
            Playlist 2
          </div>
          <div
            class="songs-count"
          >
            0
            ㆍSongs
          </div>
        </div>
        <div
          class="playlist-enter"
        >
          <img
            alt="playlistEnter"
            src="plEnter.png"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`PlaylistSet renders PlaylistSet component 5`] = `
<button
  style="font-size: 20px; height: 70%; width: 35%;"
>
  Delete Playlist
</button>
`;
